### DISCUSSION TITLE
AnyOf or OneOf validation in adonisJs

### DISCUSSION THREAD

**Discussion by Vishwas-verma** [adonisjs/core#2771] [General]
Hye,
is there a way to choose between two keys while validating , like we have **anyOf** or **oneOf** in Json schema is there similar thing in adonisJS.
**Here is ny snippet**

```
  public schema = schema.create({
    phone_number: schema.string.optional({}, [rules.unique({table: 'users', column: 'phone_number'})]),
    otp: schema.number.optional([rules.requiredIfExists('phone_number')]),
    email: schema.string.optional({}, [rules.unique({table: 'users', column: 'email'})]),
    password: schema.string.optional({},[rules.requiredIfExists('email')])
  })
```

My use case is either user enter phone_number and otp or else enter email and password , not all at same time.

@thetutlage is it possible in adonis ?

**Comment by Vishwas-verma** [comment]
> Hye,
> 
> is there a way to choose between two keys while validating , like we have **anyOf** or **oneOf** in Json schema is there similar thing in adonisJS.
> 
> **Here is ny snippet**
> 
> 
> 
> ```
> 
>   public schema = schema.create({
> 
>     phone_number: schema.string.optional({}, [rules.unique({table: 'users', column: 'phone_number'})]),
> 
>     otp: schema.number.optional([rules.requiredIfExists('phone_number')]),
> 
>     email: schema.string.optional({}, [rules.unique({table: 'users', column: 'email'})]),
> 
>     password: schema.string.optional({},[rules.requiredIfExists('email')])
> 
>   })
> 
> ```
> 
> 
> 
> My use case is either user enter phone_number and otp or else enter email and password , not all at same time.
> 
> 
> 
> 
> 
> @thetutlage is it possible in adonis ? 
> 
Any update anyone ??

**Reply by thetutlage** [reply]
I think we need a `requireIfNotExists` rule for this and the schema can look as follows:

```ts
schema.create({
  phone_number: schema.string({}, [
    requireIfNotExists('email')
  ]),
  otp: schema.string({}, [
    requireIfExist('phone_number')
  ]),
  email: schema.string({}, [
    requireIfNotExists('phone_number')
  ]),
  password: schema.string({}, [
    requireIfExist('email')
  ]),
})
```

Feel free to send a PR to add this rule. I will get it merge :)

**Reply by Vishwas-verma** [reply]
Hye @thetutlage I will send you the PR for this.
However this will create a little confusion when dealing with large schema's preferably conditional schema's.
What I was asking is some thing like OneOf or AnyOf in JSON schema,that will make it easier to write complex conditional schemas . 
Though this will work but that will help a lot in writing code efficiently and will be easily
understandable. 😃
